File: python/defaults/instance/routing.py
#
# Prudence Routing
#

from java.io import File
from com.threecrickets.prudence.util import IoUtil

# Hosts

execute_or_default('instance/hosts/')

# Applications

applications = component.context.attributes['com.threecrickets.prudence.applications'] = []
applications_dir = File(document.source.basePath, 'applications')

properties_file = File(applications_dir, 'applications.properties')
properties = IoUtil.loadProperties(properties_file)
save_properties = False
application_files = applications_dir.listFiles()
for application_file in application_files:
    last_modified = str(application_file.lastModified())
    if not application_file.directory and application_file.name[-4:] == '.zip' and properties.getProperty(application_file.name, '') != last_modified:
        print 'Unpacking "' + application_file.name + '"...'
        IoUtil.unzip(application_file, applications_dir)
        properties.setProperty(application_file.name, last_modified)
        save_properties = True

if save_properties:
    IoUtil.saveProperties(properties, properties_file)

application_dirs = applications_dir.listFiles()
for application_dir in application_dirs:
    if application_dir.isDirectory():
        application_name = application_dir.name
        application_internal_name = application_dir.name;
        application_logger_name = application_dir.name
        application_base_path = application_dir.path
        application_default_url = '/' + application_dir.name + '/'
        application_base = 'applications/' + application_dir.name + '/'
        execute_or_default(application_base, 'defaults/application/')
        applications.append(application_instance)

if len(applications) == 0:
    print 'No applications found. Exiting.'
    System.exit(0)

File: javascript/defaults/instance/routing.js
//
// Prudence Routing
//
// Copyright 2009-2011 Three Crickets LLC.
//
// The contents of this file are subject to the terms of the LGPL version 3.0:
// http://www.opensource.org/licenses/lgpl-3.0.html
//
// Alternatively, you can obtain a royalty free commercial license with less
// limitations, transferable or non-transferable, directly from Three Crickets
// at http://threecrickets.com/
//

importClass(
	java.io.File,
	java.util.ArrayList,
	com.threecrickets.prudence.util.IoUtil)

// Hosts

executeOrDefault('instance/hosts/')

// Applications

var applications = new ArrayList()
component.context.attributes.put('com.threecrickets.prudence.applications', applications)
var applicationsDir = new File(document.source.basePath, 'applications')

var propertiesFile = new File(applicationsDir, 'applications.properties')
var properties = IoUtil.loadProperties(propertiesFile)
var saveProperties = false
var applicationFiles = applicationsDir.listFiles()
for(var i in applicationFiles) {
	var applicationFile = applicationFiles[i]
	if(!applicationFile.directory && applicationFile.name.endsWith('.zip') && properties.getProperty(applicationFile.name, '') != applicationFile.lastModified()) {
		print('Unpacking "' + applicationFile.name + '"...\n')
		IoUtil.unzip(applicationFile, applicationsDir)
		properties.setProperty(applicationFile.name, applicationFile.lastModified())
		saveProperties = true
	}
}
if(saveProperties) {
	IoUtil.saveProperties(properties, propertiesFile)
}

var applicationDirs = applicationsDir.listFiles()
for(var i in applicationDirs) {
	var applicationDir = applicationDirs[i]
	if(applicationDir.directory && !applicationDir.hidden) {
		var applicationName = applicationDir.name
		var applicationInternalName = applicationDir.name
		var applicationLoggerName = applicationDir.name
		var applicationBasePath = applicationDir.path
		var applicationDefaultURL = '/' + applicationDir.name + '/'
		var applicationBase = 'applications/' + applicationDir.name + '/'
		executeOrDefault(applicationBase, 'defaults/application/')
		applications.add(applicationInstance)
	}
}

if(applications.length == 0) {
	print('No applications found. Exiting.\n')
	System.exit(0)
}
